src/librustc_mir/borrow_check.rs

```diff
@@ -1596,26 +1621,36 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
 
                 match proj.elem {
                     ProjectionElem::Deref => {
-                        if autoderef {
-                            self.append_lvalue_to_string(&proj.base, buf, Some(autoderef));
+                        if let Some(field) = self.is_upvar_field_projection(&proj.base) {
+                            let var_index = field.index();
+                            let name = self.mir.upvar_decls[var_index].debug_name.to_string();
+                            if self.mir.upvar_decls[var_index].by_ref {
+                                buf.push_str(&name);
+                            } else {
+                                buf.push_str( &format!("*{}", &name));
+                            }
                         } else {
-                            buf.push_str(&"(*");
-                            self.append_lvalue_to_string(&proj.base, buf, Some(autoderef));
-                            buf.push_str(&")");
+                            if autoderef {
+                                self.append_lvalue_to_string(&proj.base, buf, Some(autoderef));
+                            } else {
+                                buf.push_str(&"(*");
```

Comment:
It's a separate bug (https://github.com/rust-lang/rust/issues/45185), but I'm pretty sure the `()` here are not necessary. That said, there is a fix for this in https://github.com/rust-lang/rust/pull/46040 (which will obviously conflict with this PR).